satpy/readers/seviri_l1b_hrit.py

```diff
@@ -751,49 +748,30 @@ def pad_hrv_data(self, res):
     def calibrate(self, data, calibration):
         """Calibrate the data."""
         tic = datetime.now()
-        channel_name = self.channel_name
-
-        if calibration == 'counts':
-            res = data
-        elif calibration in ['radiance', 'reflectance', 'brightness_temperature']:
-            # Choose calibration coefficients
-            # a) Internal: Nominal or GSICS?
-            band_idx = self.mda['spectral_channel_id'] - 1
-            if self.calib_mode != 'GSICS' or self.channel_name in VIS_CHANNELS:
-                # you cant apply GSICS values to the VIS channels
-                coefs = self.prologue["RadiometricProcessing"]["Level15ImageCalibration"]
-                int_gain = coefs['CalSlope'][band_idx]
-                int_offset = coefs['CalOffset'][band_idx]
-            else:
-                coefs = self.prologue["RadiometricProcessing"]['MPEFCalFeedback']
-                int_gain = coefs['GSICSCalCoeff'][band_idx]
-                int_offset = coefs['GSICSOffsetCount'][band_idx] * int_gain
-
-            # b) Internal or external? External takes precedence.
-            gain = self.ext_calib_coefs.get(self.channel_name, {}).get('gain', int_gain)
-            offset = self.ext_calib_coefs.get(self.channel_name, {}).get('offset', int_offset)
-
-            # Convert to radiance
-            data = data.where(data > 0)
-            res = self._convert_to_radiance(data.astype(np.float32), gain, offset)
-            line_mask = self.mda['image_segment_line_quality']['line_validity'] >= 2
-            line_mask &= self.mda['image_segment_line_quality']['line_validity'] <= 3
-            line_mask &= self.mda['image_segment_line_quality']['line_radiometric_quality'] == 4
-            line_mask &= self.mda['image_segment_line_quality']['line_geometric_quality'] == 4
-            res *= np.choose(line_mask, [1, np.nan])[:, np.newaxis].astype(np.float32)
-
-        if calibration == 'reflectance':
-            solar_irradiance = CALIB[self.platform_id][channel_name]["F"]
-            res = self._vis_calibrate(res, solar_irradiance)
-
-        elif calibration == 'brightness_temperature':
-            cal_type = self.prologue['ImageDescription'][
-                'Level15ImageProduction']['PlannedChanProcessing'][self.mda['spectral_channel_id']]
-            res = self._ir_calibrate(res, channel_name, cal_type)
-
+        calib = SEVIRICalibrationHandler(
+            platform_id=self.platform_id,
+            channel_name=self.channel_name,
+            coefs=self._get_calib_coefs(self.channel_name),
+            calib_mode=self.calib_mode,
+            scan_time=self.start_time
+        )
+        res = calib.calibrate(data, calibration)
+        if calibration in ['radiance', 'reflectance', 'brightness_temperature']:
+            res = self._mask_bad_quality(res)
```

Comment:
Can this be made optional? Sometimes it's useful to have the bad quality data included.

Reply:
I agree, but I would like to keep the focus on calibration here and do that in a separate PR. I'll create an issue.